addons/mrp_subcontracting/models/stock_picking.py

```diff
@@ -18,96 +19,107 @@ def _compute_display_action_record_components(self):
                 continue
             if not picking._is_subcontract():
                 continue
-            # Hide if no move is tracked
+            # Hide if no components are track
             subcontracted_productions = picking._get_subcontracted_productions()
             subcontracted_moves = subcontracted_productions.mapped('move_raw_ids')
-            subcontracted_moves |= subcontracted_productions.mapped('move_finished_ids')
             if all(subcontracted_move.has_tracking == 'none' for subcontracted_move in subcontracted_moves):
                 continue
             # Hide if the production is to close
             if not subcontracted_productions.filtered(lambda mo: mo.state not in ('to_close', 'done')):
                 continue
             picking.display_action_record_components = True
 
-    @api.depends('state')
-    def _compute_display_view_subcontracted_move_lines(self):
-        for picking in self:
-            # Hide if not encoding state
-            if picking.state in ('draft', 'cancel'):
-                continue
-            if not picking._is_subcontract():
-                continue
-            # Hide until state done if no move is tracked, if tracked until something was produced
-            subcontracted_productions = picking._get_subcontracted_productions()
-            subcontracted_moves = subcontracted_productions.mapped('move_raw_ids')
-            subcontracted_moves |= subcontracted_productions.mapped('move_finished_ids')
-            if all(subcontracted_move.has_tracking == 'none' for subcontracted_move in subcontracted_moves):
-                if picking.state != 'done':
-                    continue
-            # Hide if nothing was produced
-            if all(subcontracted_move.quantity_done == 0 for subcontracted_move in subcontracted_moves):
-                continue
-            picking.display_view_subcontracted_move_lines = True
-
     # -------------------------------------------------------------------------
     # Action methods
     # -------------------------------------------------------------------------
+
+    def action_cancel(self):
+        for picking in self:
+            picking._get_subcontracted_productions()._action_cancel()
+        return super(StockPicking, self).action_cancel()
+
     def action_done(self):
+        res = super(StockPicking, self.with_context(skip_mo=True)).action_done()
```

Comment:
you need to explain the skip_mo context key (dont_create_subcontract_order)

maybe split self in two recordset: subcontract and not subcontract

``` python
is_subcontract = {picking: any(move.is_subcontract for move in picking.move_lines) for picking in self}
for picking in self:
    if not is_subcontract.get(picking):
        continue
   production = move.move_orig_ids.production_id
```